### PR TITLE
✨ [WIP]Feature

### PR DESCRIPTION
**Pull request by Adhikram**
Update Count & call by Username feature

[Notion Task Link](https://www.notion.so/User-profile-Cleanup-7dc6fc7a16d744ffb00701022453dfd2)

### CODE REVIEW

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 1]
```diff
@@ -119,6 +121,35 @@ def update_college
...
+          end
+          user.save
+        end
+      end
+
+      def get_by_username
+        unless !!params['id'].match(/^\d{1,99}$/)
+          # dodging teh params/id call
+          # return render_error({ message: 'No user data found' }) unless !!params['id'].match(/^(?!.*\.\.)(?!.*\.$)[^\W][\w.]{4,29}$/)
+
```
Comment (sam0000-cmd) [inline devsnest/app/controllers/api/v1/users_controller.rb:145]: remove comments

File: devsnest/app/resources/api/v1/user_resource.rb [hunk 2]
```diff
@@ -4,7 +4,7 @@ module Api
   module V1
     class UserResource < JSONAPI::Resource
       attributes :email, :name, :password, :discord_id, :web_active, :username, :score, :discord_active, :batch, :grad_status, :grad_specialization, :grad_year, :grad_start, :grad_end,
-                 :github_url, :linkedin_url, :resume_url, :dob, :registration_num, :college_id, :image_url, :google_id, :bot_token
+                 :github_url, :linkedin_url, :resume_url, :dob, :registration_num, :college_id, :image_url, :google_id, :bot_token,:update_count
```
Comment (kaydee0502) [inline devsnest/app/resources/api/v1/user_resource.rb:7]: space after coma

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -119,6 +121,35 @@ def update_college
 
         params['data']['attributes'].delete 'college_name'
       end
+
+      def update_username
+        # continue if !!params['data']['id'].match(/^\d{1,99}$/)
+        id = params['data']['id']
+        unless User.find_by(id: id).username == params['data']['attributes']['username']
+          user = User.find_by(id: id)
```
Comment (sam0000-cmd) [inline devsnest/app/controllers/api/v1/users_controller.rb:129]: Merge these two db calls

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 4]
```diff
@@ -119,6 +121,35 @@ def update_college
...
+          return render_error({ message: 'username format missmatched' }) unless !!params['data']['attributes']['username'].match(/^(?!.*\.\.)(?!.*\.$)[^\W][\w.]{4,29}$/)
+
+          if user.update_count >= 10
+            params['data']['attributes']['username'] = user.username
+          else
+            user.update_count = user.update_count + 1
```
Comment (sam0000-cmd) [inline devsnest/app/controllers/api/v1/users_controller.rb:135]: user.update(update_count:update_count+1)

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 5]
```diff
@@ -119,6 +121,35 @@ def update_college
...
+      end
+
+      def get_by_username
+        unless !!params['id'].match(/^\d{1,99}$/)
+          # dodging teh params/id call
+          # return render_error({ message: 'No user data found' }) unless !!params['id'].match(/^(?!.*\.\.)(?!.*\.$)[^\W][\w.]{4,29}$/)
+
+          username = params[:id]
+          user = User.find_by(username: username)
+          return render_not_found unless user.present?
```
Comment (sam0000-cmd) [inline devsnest/app/controllers/api/v1/users_controller.rb:148]: user = User.find_by(username: params[:id])

File: devsnest/app/models/user.rb [hunk 6]
```diff
@@ -6,6 +6,11 @@ class User < ApplicationRecord
          jwt_revocation_strategy: JwtBlacklist
   after_create :create_bot_token
   enum user_type: %i[user admin]
+  after_create :create_username
+  def create_username
+    curr_name = email.split('@')[0]
+    update_attribute(:username, curr_name)
+  end
```
Comment (sam0000-cmd) [inline devsnest/app/models/user.rb:13]: update_attribute(:username, email.split('@')[0])

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 5 above)
Comment (Adhikram) [reply devsnest/app/controllers/api/v1/users_controller.rb:148]: Done

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 4 above)
Comment (Adhikram) [reply devsnest/app/controllers/api/v1/users_controller.rb:135]: Fixed by  params['data']['attributes']['update_count'] = user.update_count + 1

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 3 above)
Comment (Adhikram) [reply devsnest/app/controllers/api/v1/users_controller.rb:129]: Fixed By : user = User.find_by(id: id)

File: devsnest/app/resources/api/v1/user_resource.rb (hunk 2 above)
Comment (Adhikram) [reply devsnest/app/resources/api/v1/user_resource.rb:7]: FIXED

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 1 above)
Comment (Adhikram) [reply devsnest/app/controllers/api/v1/users_controller.rb:145]: Removed

File: devsnest/app/models/user.rb (hunk 6 above)
Comment (Adhikram) [reply devsnest/app/models/user.rb:13]: Done

**Review by sam0000-cmd — APPROVED** [review]

File: devsnest/app/models/user.rb [hunk 7]
```diff
@@ -6,6 +6,10 @@ class User < ApplicationRecord
          jwt_revocation_strategy: JwtBlacklist
   after_create :create_bot_token
   enum user_type: %i[user admin]
+  after_create :create_username
```
Comment (vedansh) [inline devsnest/app/models/user.rb:9]: add a new line

**Comment by vedansh** [comment]
@Adhikram did you forgot to push routes file

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 8]
```diff
@@ -119,6 +119,30 @@ def update_college
...
+            params['data']['attributes']['username'] = user.username
+          else
+            params['data']['attributes']['update_count'] = user.update_count + 1
+          end
+
+        end
+      end
+
+      def get_by_username
+        unless !!params['id'].match(/^\d{1,99}$/)
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:139]: unless and then !!, can we simplify this

Comment (vedansh) [reply devsnest/app/controllers/api/v1/users_controller.rb:139]: no, this breaks the general flow, can you create one more function which redirects to /user/:id

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 9]
```diff
@@ -119,6 +119,30 @@ def update_college
 
         params['data']['attributes'].delete 'college_name'
       end
+
+      def update_username
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:123]: simplify, too complicated

can we have an early return to this function, if username filed is not present or same as old username

Then validation check for username, can we create a function in User and call it here

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 10]
```diff
@@ -119,6 +119,30 @@ def update_college
...
+        id = params['data']['id']
+        user = User.find_by(id: id)
+        unless user.username == params['data']['attributes']['username']
+          return render_error({ message: 'username format missmatched' }) unless !!params['data']['attributes']['username'].match(/^(?!.*\.\.)(?!.*\.$)[^\W][\w.]{4,29}$/)
+
+          if user.update_count >= 2
+            params['data']['attributes']['username'] = user.username
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:130]: We should never fail silently, return an error is this is changed

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 11]
```diff
@@ -119,6 +119,32 @@ def update_college
 
         params['data']['attributes'].delete 'college_name'
       end
+
+      def update_username
+        id = params['data']['id']
+        user = User.find_by(id: id)
+        return render_error({ message: 'username format missmatched' }) unless !!params['data']['attributes']['username']
```
Comment (Adhikram) [inline devsnest/app/controllers/api/v1/users_controller.rb:126]: Disable Update of other users

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 12]
```diff
@@ -119,6 +125,21 @@ def update_college
 
         params['data']['attributes'].delete 'college_name'
       end
+      
+      def update_username
+        # byebug
+        return render_error({ message: 'Unauthorized request' }) unless context[:user].id == params['data']['id'].to_i
+        return render_error({ message: 'username format missmatched' }) unless check_username(params['data']['attributes']['username'])
+        unless context[:user].username == params['data']['attributes']['username']
```
Comment (Adhikram) [inline devsnest/app/controllers/api/v1/users_controller.rb:133]: return true if parm.. .nil? ||  context[:user].username == params['data']['attributes']['username']

File: devsnest/app/controllers/api/v1/users_controller.rb [hunk 13]
```diff
@@ -119,6 +126,19 @@ def update_college
...
       end
+
+      def update_username
+        return render_unauthorized unless @current_user.id.to_s == params['data']['id']
+        return render_error({ message: 'Username format missmatched' }) unless check_username(params['data']['attributes']['username'])
+
+        return true if params['data']['attributes']['username'].nil? || context[:user].username == params['data']['attributes']['username']
+
+        if context[:user].update_count >= 2
+          render_error({ message: 'Update count Exceeded for username' })
```
Comment (vedansh) [inline devsnest/app/controllers/api/v1/users_controller.rb:137]: do return render_error

**Review by vedansh — APPROVED** [review]

File: devsnest/app/controllers/api/v1/users_controller.rb (hunk 13 above)
Comment (vedansh) [reply devsnest/app/controllers/api/v1/users_controller.rb:137]: 10?

Comment (Adhikram) [reply devsnest/app/controllers/api/v1/users_controller.rb:137]: It was for testing purpose

**Review by LakshitXD — APPROVED** [review]

**Pull request merged**